toolComponent change

michaelcfanning opened the pull request:
@lgolding can you sanity check the schema change for the new toolComponent object?

in re: https://github.com/oasis-tcs/sarif-spec/issues/179

lgolding commented on src/Sarif/Schemata/sarif-schema.json:
```diff
@@ -1507,6 +1507,12 @@
+        "extensionIndex": {
+          "description": "The index within the run.tool.extensions array of the tool component object which describes the plug-in or tool extension that produced the result.",
+          "type": "integer",
+          "minimum": -1
```
>minimum [](start = 11, length = 7)

and: `"default": -1`

michaelcfanning replied:
Yes, did some clean up for minimum/default, please take a look at update.

---
In reply to: [255688316](https://github.com/Microsoft/sarif-sdk/pull/1277#discussion_r255688316) [](ancestors = 255688316)

lgolding commented on src/Sarif/Schemata/sarif-schema.json:
```diff
@@ -2013,51 +2019,78 @@
+        },
+
+        "extensions": {
+          "description": "Tool extensions that contributed to or reconfigured the analysis tool run.",
```
>run [](start = 96, length = 3)

"run" => "that was run" (for parallelism with description of "driver", and also I think it's clearer).

lgolding commented on src/Sarif/Schemata/sarif-schema.json:
```diff
@@ -2013,51 +2019,78 @@
 
         "name": {
-          "description": "The name of the tool.",
+          "description": "The name of the tool driver or extension.",
```
>driver or extension [](start = 47, length = 19)

How about "component" since that's what you use in the rest of the properties below.

lgolding commented on src/Sarif/Schemata/sarif-schema.json:
```diff
@@ -2013,51 +2022,78 @@
+        }
+      },
+
+      "required": [ "driver" ]
```
>"required": [ "driver" ] [](start = 6, length = 24)

Yes.

lgolding commented on src/Sarif/Schemata/sarif-schema.json:
```diff
@@ -2013,51 +2022,78 @@
+          }
+        },
+
+
```
Extra blank line.

lgolding commented:
Looks fine. I agree you should add `result.toolComponentIndex`, default -1 (meaning it came from the driver, not an additional component).

lgolding commented:
Michael, didn't we talk about linking from a `toolComponent` to a `file` object, so we could have information like file hashes? (So: `toolComponent.fileIndex`). Less important, but I still would like it: an additional `role` value of `toolComponent`.

michaelcfanning commented:
i put this in resultProvenance.extensionIndex. You don't populate this at all in cases where it came from the driver. Seems simpler for happy path production. ?

---
In reply to: [462492896](https://github.com/Microsoft/sarif-sdk/pull/1277#issuecomment-462492896) [](ancestors = 462492896)

michaelcfanning commented:
yes! on it. ok, role, yes.


---
In reply to: [462493928](https://github.com/Microsoft/sarif-sdk/pull/1277#issuecomment-462493928) [](ancestors = 462493928)

lgolding commented:
- Re: The default value: This is one of those cases where we need the default to be -1 for the sake of the serializer, because 0 is a valid value. If 0 were the default, the serializer would not emit 0, and it would like like it was absent. The serializer will refrain from emitting -1, so it will indeed be absent.
- Re: location of extensionIndex: I think it belongs in `result`. Otherwise, if your tool has rule plug-ins, you _have_ to instantiate a `provenance` object _just to hold the extension index_ -- even if you would otherwise not have had to bother.

---
In reply to: [462495693](https://github.com/Microsoft/sarif-sdk/pull/1277#issuecomment-462495693) [](ancestors = 462495693,462492896)

lgolding commented on src/Sarif/Schemata/sarif-schema.json:
```diff
@@ -518,7 +520,7 @@
           "description": "A string which indirectly specifies the absolute URI with respect to which a relative URI in the \"uri\" property is interpreted.",
           "type": "string"
-        },
+        },`
```
>` [](start = 10, length = 1)

typo

michaelcfanning commented:
the default is -1, do you see a problem i need to fix?
yes, you need to instantiate the provenance object, just as you have to if you specify invocationIndex. Can you explain why extensionIndex would be on result but invocationIndex would not? in general, we introduced the provenance object to provide option, unusual origin data for results. associating a specific result with a custom plug-in qualifies, in my opinion. 

do you have a different set of standards for what belongs in provenance?

---
In reply to: [462497387](https://github.com/Microsoft/sarif-sdk/pull/1277#issuecomment-462497387) [](ancestors = 462497387,462495693,462492896)

lgolding commented:
No problem on the default. I misunderstood you.

---
In reply to: [462498159](https://github.com/Microsoft/sarif-sdk/pull/1277#issuecomment-462498159) [](ancestors = 462498159,462497387,462495693,462492896)

lgolding commented on src/Sarif/Core/Tool.cs:
```diff
@@ -15,50 +16,46 @@ namespace Microsoft.CodeAnalysis.Sarif
+        private static IDictionary<string, SerializedPropertyInfo> CreatePropertiesFromFileVersionInfo(FileVersionInfo fileVersion)
         {
-            return this.NotificationsMetadata.HasAtLeastOneNonNullValue();
+            var toolComponent = new ToolComponent();
```
>ToolComponent [](start = 36, length = 13)

Can you create a property bag directly? Do you need a `toolComponent` to hold it?

michaelcfanning replied:
It is cumbersome to instantiate IDictionary<string, SerializedPropertyInfo> directly (I do it elsewhere in this change). This is the easier mechanism because you can use the SetProperty helper. Do you have other advice on creation? This specific changes retains the previous code pattern...

---
In reply to: [256107576](https://github.com/Microsoft/sarif-sdk/pull/1277#discussion_r256107576) [](ancestors = 256107576)

lgolding commented on src/Sarif/Visitors/InsertOptionalDataVisitor.cs:
```diff
@@ -177,7 +177,7 @@ public override Message VisitMessage(Message node)
 
                 if (node.Text == null &&
-                    _run.Tool?.GlobalMessageStrings?.TryGetValue(node.MessageId, out formatString) == true)
+                    _run.Tool.Driver.GlobalMessageStrings?.TryGetValue(node.MessageId, out formatString) == true)
```
>. [](start = 29, length = 1)

Why did you remove the null conditional operator?

michaelcfanning replied:
check line 168, where it didn't exist. Tool.Driver is a required object. we could put an arg check at the top of the method

---
In reply to: [256108174](https://github.com/Microsoft/sarif-sdk/pull/1277#discussion_r256108174) [](ancestors = 256108174)

lgolding commented on src/Sarif.Converters/SemmleQlConverter.cs:
```diff
@@ -36,6 +36,8 @@ private enum FieldIndex
         private List<Notification> _toolNotifications;
 
+        public override string ToolName { get { return "Semmle QL"; } }
+
```
In all the other converters you just wrote
```
public override string ToolName => "Semmle QL";
```

lgolding commented on src/Sarif.Converters/CppCheckConverter.cs:
```diff
@@ -20,6 +20,8 @@ public CppCheckConverter()
             _strings = new CppCheckStrings(_nameTable);
         }
 
+        public override string ToolName => ToolFormat.CppCheck;
```
>CppCheck [](start = 54, length = 8)

I know we discussed, and I think we agreed, not to assume that the `ToolFormat` would also make a good tool name. In most of the other converters, you're just hard-coding a string here (`"CppCheck"`).

michaelcfanning replied:
in fact, most converters still use ToolFormat. I'm not going to invest more here, I spent hours on this change DRYing out test code, removing embedded literals. We can revisit this in the future if need be.

---
In reply to: [256111009](https://github.com/Microsoft/sarif-sdk/pull/1277#discussion_r256111009) [](ancestors = 256111009)

lgolding commented on src/Sarif.Converters.UnitTests/ConverterTestsBase.cs:
```diff
@@ -43,5 +44,36 @@ public SarifLog RunTestCase(string inputData, string expectedResult, bool pretty
             return log;
         }
+
+        private static SarifLog BuildToolSpecificEmptyLog()
```
Nice.

lgolding commented on src/Sarif.FunctionalTests/PropertyBagConverterTests.cs:
```diff
@@ -14,34 +15,49 @@ public class PropertyBagConverterTests
+            int intPropertyValue = 42;
+
+            string stringPropertyName = nameof(stringPropertyName);
+            string stringPropetyValue = "'\"\\'";
```
>stringPropetyValue [](start = 19, length = 18)

`Propety` => `Property`

lgolding commented on src/Sarif.UnitTests/Readers/IsSuppressedInSourceConverterTests.cs:
```diff
@@ -12,28 +13,13 @@ public class InSourceSuppressionConverterTests : JsonTests
+        {            
+            string expected = CreateCurrentV2SarifLogText(
+                resultCount: 1,
+                (log) => {
```
>log [](start = 17, length = 3)

I absolutely love this pattern.

lgolding commented on src/Sarif.UnitTests/TestData/PrereleaseCompatibilityTransformer/ExpectedOutputs/ComprehensiveToolProperties.sarif:
```diff
@@ -0,0 +1,63 @@
+          "properties": {
+            "tags": [
+  "singleTag"
+],
```
formatting.

lgolding commented on src/Sarif.UnitTests/TestData/PrereleaseCompatibilityTransformer/Inputs/ComprehensiveToolProperties.sarif:
```diff
@@ -0,0 +1,60 @@
+          "tags": [ "singleTag" ],
+          "propertyKey": "propertyValue"
+        },
+        "globalMessageStrings": {
```
>globalMessageStrings [](start = 9, length = 20)

In 11-28, these three properties didn't yet live in `tool`, right? Global message strings and rules metadata lived in `run.resources`, and notification metadata didn't yet exist.

lgolding commented on src/Sarif.UnitTests/Writers/ResultLogJsonWriterTests.cs:
```diff
@@ -330,95 +288,22 @@ public void ResultLogJsonWriter_CannotWriteInvocationTwice()
+                    {
+                        new Invocation
+                        {
+                            ConfigurationNotifications = new List<Notification>(s_notifications)
```
>s_notifications [](start = 80, length = 15)

Hah. Really nice.

michaelcfanning replied:
All the pieces were here. :)

---
In reply to: [256121795](https://github.com/Microsoft/sarif-sdk/pull/1277#discussion_r256121795) [](ancestors = 256121795)

lgolding commented:
>             string version = forceUpdate ? "2.0.0" : (string)sarifLog["version"];

Why is this no longer needed? If it's not needed, the comment is obsolete.

---
Refers to: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs:42 in 2dc4f02. [](commit_id = 2dc4f02bb30b46a98ea694d27357c35972e84b9a, deletion_comment = True)

michaelcfanning commented:
>             string version = forceUpdate ? "2.0.0" : (string)sarifLog["version"];

This force argument always existed because some tests lie about their version: i.e., the profess to have a format that they don't. I cleaned that up and removed this arg. I'll remove the dead comment in the next PR. also, please note: the modified log arg can go away now. at this point, any transformation of a prerelease format entails a change (because a required property, tool, has been updated).

---
In reply to: [462935363](https://github.com/Microsoft/sarif-sdk/pull/1277#issuecomment-462935363) [](ancestors = 462935363)

---
Refers to: src/Sarif/Writers/PrereleaseCompatibilityTransformer.cs:42 in 2dc4f02. [](commit_id = 2dc4f02bb30b46a98ea694d27357c35972e84b9a, deletion_comment = True)
